fix(ui): Remove the ability to change namespaces via the UI in Managed Namespace Mode. Closes #5577

Radolumbo opened the pull request:
addresses https://github.com/argoproj/argo-workflows/issues/5577

In older versions of Argo, running the server with a managed namespace prevented you from being able to browse to other namespaces within the UI. This ability was lost in V3, such that you could manually change the URL to attempt to view resources in other namespaces. Additionally, in the older versions, Simply browsing to "/workflows" would show you workflows in the managed namespace, so the users did not even really need to be aware of where their workflows were running at all.

It seems like this behavior is still INTENDED in v3, given code in the [namespace filter](https://github.com/argoproj/argo-workflows/blob/master/ui/src/app/shared/components/namespace-filter.tsx#L6) and in the [app router](https://github.com/argoproj/argo-workflows/blob/master/ui/src/app/app-router.tsx#L60-L66). However, the actual behavior differs.

I followed a similar paradigm to V2, where each page was responsible for checking whether you were running in a managed namespace. For example, see the code in [workflows list](https://github.com/argoproj/argo-workflows/blob/release-2.10/ui/src/app/workflows/components/workflows-list/workflows-list.tsx#L194-L201) or in [workflow templates list](https://github.com/argoproj/argo-workflows/blob/release-2.10/ui/src/app/workflow-templates/components/workflow-template-list/workflow-template-list.tsx#L105-L109).

This should not affect the behavior of the UI outside of managed namespace mode with one notable exception: in workflows-list, the saveHistory call was not properly setting the URL to include the filter params due to the order of operations between "||" and "+". This has been fixed.

Checklist:

* [x] My organization is added to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).

<!--

Please leave your PR in draft if you don't need a review yet. 

To fix failing `CI / Codegen`, run `make codegen`. 

-->

alexec commented on ui/src/app/app-router.tsx:
```diff
@@ -79,12 +79,12 @@ export const AppRouter = ({popupManager, history, notificationsManager}: {popupM
                             {
                                 title: 'Workflows',
-                                path: workflowsUrl + '/' + namespace,
+                                path: workflowsUrl + (Utils.managedNamespace ? '' : '/' + namespace),
```
I think you can simplify this to `Utils.getNamespace(namespace)`?

Radolumbo replied:
So, this is actually blanking out the namespace entirely. The goal is to just have it point to /workflows rather than /workflows/namespace. I think, ideally, in managed namespace mode, you don't even need to be aware of the NS. This is how it went in V2, as well.

What say you?

alexec replied:
I think we should keep the long URLs, they should work for deep-linking:

`/workflows/not-found`

Should result in an error if your namespace is not found.

But perhaps in managed namespace mode:

`/workflows`

Should only load from the managed namespace, rather that attempt/fail to redirect

Radolumbo replied:
Yeah, that's what my PR does. It overrides the state within each page to point to the managed namespace, so that browsing to

`/workflows`

Only loads from the managed namespace. Thus why I'm having the navbar point just to the root URL for each page, again only in managed mode.

With these changes, when in managed mode, an attempt to browse to something like `/workflows/other_namespace` will redirect you back to `/workflows`. 

In unmanaged mode, the same behavior of not-found namespaces remains.

Radolumbo commented:
https://github.com/argoproj/argo-workflows/pull/5729/checks?check_run_id=2401672970

Unclear to me what this failure means. It seems like it may have been a timeout/temporary failure. Is there a way to just rerun it without adding commits?

alexec commented on ui/src/app/app-router.tsx:
```diff
@@ -79,12 +79,12 @@ export const AppRouter = ({popupManager, history, notificationsManager}: {popupM
                             {
                                 title: 'Workflows',
-                                path: workflowsUrl + '/' + namespace,
+                                path: workflowsUrl + (Utils.managedNamespace ? '' : '/' + namespace),
```
I think you can simplify this to `Utils.getNamespace(namespace)`?

Radolumbo replied:
So, this is actually blanking out the namespace entirely. The goal is to just have it point to /workflows rather than /workflows/namespace. I think, ideally, in managed namespace mode, you don't even need to be aware of the NS. This is how it went in V2, as well.

What say you?

alexec commented:
Looks good. Can I request you attach some screenshots or other evidence of testing? You'd be amazed at the number of untested PRs we get.

Radolumbo commented:
I don't have Events installed so you'll see a couple "resource not founds" but mostly you can look at how the URLs are rendered since that's ultimately what's being changed here (I see the exact same errors using the latest argo server version as well). You'll see the namespace statically set to "argo1", as well.

Showing the behavior in namespaced mode. You'll see it doesn't keep the namespace in the path and only serves the resources in the managed namespace no matter what (note that for some of the less recently updated pages, due to how they're written, it's still part of the query params):
![namespaced](https://user-images.githubusercontent.com/3118655/115632408-a4b52200-a2d5-11eb-9946-c5749ef2fbd9.gif)

And to show the behavior hasn't changed in "normal" mode. You can see that when there is a namespace in the state or when there isn't, the pages respect whichever is the current state when switching between them:
![normal](https://user-images.githubusercontent.com/3118655/115733725-24cf9c00-a357-11eb-8076-ebcf11eb51ab.gif)

alexec commented on ui/src/app/app-router.tsx:
```diff
@@ -79,12 +79,12 @@ export const AppRouter = ({popupManager, history, notificationsManager}: {popupM
                             {
                                 title: 'Workflow Templates',
-                                path: workflowTemplatesUrl + '/' + namespace,
+                                path: workflowTemplatesUrl + (Utils.managedNamespace ? '' : '/' + namespace),
```
could this repeated piece of code be refactored please?

```
const suffix = Utils.managedNamespace ? '' : '/' + namespace;
```

alexec commented on ui/src/app/archived-workflows/components/archived-workflow-list/archived-workflow-list.tsx:
```diff
@@ -47,7 +47,7 @@ export class ArchivedWorkflowList extends BasePage<RouteComponentProps<any>, Sta
         this.state = {
             pagination: {offset: this.queryParam('offset'), limit: parseLimit(this.queryParam('limit')) || savedOptions.pagination.limit},
-            namespace: this.props.match.params.namespace || '',
+            namespace: (Utils.managedNamespace ? Utils.managedNamespace : this.props.match.params.namespace) || '',
```
as above - this is copied and pasted - could it be refactored into a shared function?

Radolumbo commented:
@alexec Please see https://github.com/argoproj/argo-workflows/pull/5729/commits/1a6bc67cf1f3f887ad77e02a70b2bef42a421401 -- I refactored the code to encapsulate the behavior of "use managed namespace if set, else use input namespace" into a function. For clarity, I also renamed an existing Util function.

Radolumbo commented:
Any other thoughts/concerns?

alexec commented on ui/src/app/shared/utils.ts:
```diff
@@ -102,8 +102,13 @@ export const Utils = {
-    // return a namespace, never return null/undefined, defaults to "default"
+    // return a namespace, favoring managed namespace when set
     getNamespace(namespace: string) {
+        return this.managedNamespace || namespace;
```
I think we should remove getNamespaceWithDefault and just use this function

Radolumbo replied:
I was a little worried that I don't fully understand the implications of how that function is currently being used. It was in places where, for example, you create workflows, so I can imagine having the extra protection of at least creating it within the default namespace in the worst case is useful. Probably worth leaving, but I'll defer to you. Want me to remove it and try experimenting with the effects?

alexec replied:
um, no, lets try it out and see

Radolumbo replied:
Sorry, the use of both "no" and "let's try it out" has me confused -- did you want me to leave it as is or make the change? 😅

alexec replied:
leave it, lets test it a bit

Radolumbo commented on ui/src/app/cron-workflows/components/cron-workflow-list/cron-workflow-list.tsx:
```diff
@@ -89,7 +89,8 @@ export class CronWorkflowList extends BasePage<RouteComponentProps<any>, State>
 
     private saveHistory() {
-        this.url = uiUrl('cron-workflows/' + this.state.namespace || '');
+        const newNamespace = Utils.managedNamespace ? '' : this.state.namespace;
```
Nah, this behavior is different. This is returning blank for the path if there's a managed namespace (e.g. just cron-workflows/), else setting the path to the current namespace (e.g. cron-workflows/argo). It seemed like a bit of weird behavior to wrap up in a function, and it's not in that many places.
